cmake/templates/setup.bat.in

```diff
@@ -54,8 +54,14 @@ if NOT EXIST %_SETUP_TMP% (
   exit 1
 )
 
-REM invoke Python script to generate necessary exports of environment variables
-%_PYTHON% "%_SETUP_UTIL%" %* > %_SETUP_TMP%
+REM invoke setup script to generate necessary exports of environment variables
+set _SETUP_UTIL_EXE=@SETUP_DIR@/_setup_util.py.exe
+if EXIST "%_SETUP_UTIL_EXE%" (
+  "%_SETUP_UTIL_EXE%" %* > %_SETUP_TMP%
+) else (
```

Comment:
When would this ever be the case? I thought the wrapping executable will be installed unconditionally.

The the whole logic to determine `_PYTHON` can be removed.

Reply:
I played around with this a little bit, and found that removing the fall back route would lead to error during build (`catkin_make_isolated`), the reason is that no wrapper will be created for `devel/<package>/_setup_util.py`, and that setup script has to be executed with the Python interpreter.

only a Python script is generated for devel space: https://github.com/ros/catkin/blob/kinetic-devel/cmake/catkin_generate_environment.cmake#L40

in order to have this change, we need to change how `_setup_util.py` is exported to devel space first

Reply:
> only a Python script is generated for devel space: https://github.com/ros/catkin/blob/kinetic-devel/cmake/catkin_generate_environment.cmake#L40

It looks to me that the line you referenced just lacks the creation of the wrapper. Most of the logic of `add_python_executable` just not with an installation but writing the executable into the devel space.

Reply:
> It looks to me that the line you referenced just lacks the creation of the wrapper. Most of the logic of `add_python_executable` just not with an installation but writing the executable into the devel space.

did some more experiments, and it still wouldn't work. ultimately, I think the reason is that this [safe_execute_process](https://github.com/ros/catkin/blob/kinetic-devel/cmake/all.cmake#L207) is called before any of the win32 wrappers would get installed. Therefore, it would always fail while trying to execute `_setup_util.py.exe` under `devel\<pkg>`.

solutions I can think of:
- keep the current fall back route
- always execute `_setup_util.py` with `python` regardless of whether the wrapper exists or not

Reply:
> it still wouldn't work. ultimately, I think the reason is that this [safe_execute_process](https://github.com/ros/catkin/blob/kinetic-devel/cmake/all.cmake#L207) is called before any of the win32 wrappers would get installed.

[catkin_generate_environment](https://github.com/ros/catkin/blob/6157b75b19780278d6892364c98f8f6b4951d50f/cmake/all.cmake#L189) is being called before [safe_execute_process](https://github.com/ros/catkin/blob/6157b75b19780278d6892364c98f8f6b4951d50f/cmake/all.cmake#L207). So if you would expand the template directly into the target location in the devel space I am pretty sure it would work. As mentioned before you can't use the current function as is since you can't rely on the install to happen.

> always execute `_setup_util.py` with `python` regardless of whether the wrapper exists or no

I you prefer this way that is fine with me too.

Reply:
> So if you would expand the template directly into the target location in the devel space I am pretty sure it would work. As mentioned before you can't use the current function as is since you can't rely on the install to happen.

Just curious, what does the "template" mean here? Does it mean the newly added `add_python_executable`?

here is my understanding of this problem, please correct me if anything is not accurate:
```
${catkin_EXTRAS_DIR}/templates/_setup_util.py.in -> devel/<pkg>/_setup_util.py
|--> through atomic_configure_file(), basically configure and copy

${catkin_EXTRAS_DIR}/templates/_setup_util.py.in -> install/_setup_util.py
|--> through configure_file() then catkin_install_python()
```
given the difference between the two flows, `devel/<pkg>/_setup_util.py` is copied into `devel` instead of installed. The problem with it is that [safe_execute_process](https://github.com/ros/catkin/blob/6157b75b19780278d6892364c98f8f6b4951d50f/cmake/all.cmake#L207) in `all.cmake` will execute `generate_cache_setup.py`, and in that script, `devel/<pkg>/env.bat` (calls into `setup.bat`) gets executed. Since this happens before install happens, even if we add `add_python_executable` for `devel/<pkg>/_setup_util.py`, the wrapper will not have been created.

I am not sure how this could be changed/done, or how much longer this will take, so have reverted all changes to `setup.bat.in` first since current approach (always launch with Python interpreter) still works.

Reply:
> Just curious, what does the "template" mean here?

The `python_win32_wrapper.cpp.in` file is a template which is being expanded by `configure_file()`.

> Since this happens before install happens, even if we add `add_python_executable` for `devel/<pkg>/_setup_util.py`, the wrapper will not have been created.

Exactly, that is what I tried to describe. You can't rely on `install` even ever being triggered (e.g. when using `catkin_make` no install target is triggered). The early `env.bat` call is just another case.

Instead of adding a `add_python_executable()` call as is a subset of the logic needs to be executed which compiles the executable immediately as well as places the resulting binary into the devel space via `RUNTIME_OUTPUT_DIRECTORY`.

> [I] have reverted all changes to `setup.bat.in` first since current approach (always launch with Python interpreter) still works. 

That sounds good to me.